torch/nn/modules/loss.py

```diff
@@ -1246,6 +1253,144 @@ def forward(self, anchor: Tensor, positive: Tensor, negative: Tensor) -> Tensor:
+        reduction (string, optional): Specifies the (optional) reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed. Default: ``'mean'``
+
+
+    Shape:
+        - Input: :math:`(N, *)` where :math:`*` represents any number of additional dimensions
+            as supported by the distance function.
+        - Output: If :attr:`reduction` is ``'none'``, then a tensor of shape :math:`(N)`,
+            or a scalar otherwise.
+
+    Example::
+
+    >>> # Initialize embeddings
+    >>> embedding = nn.Embedding(1000, 128)
+    >>> anchor_ids = torch.randint(0, 1000, (1,), requires_grad=True)
+    >>> positive_ids = torch.randint(0, 1000, (1,), requires_grad=True)
+    >>> negative_ids = torch.randint(0, 1000, (1,), requires_grad=True)
+    >>> anchor = embedding(anchor_ids)
+    >>> positive = embedding(positive_ids)
+    >>> negative = embedding(negative_ids)
+    >>>
+    >>> # Built-in Distance Function
+    >>> triplet_loss = nn.TripletMarginLossWithDistance(distance_function=nn.PairwiseDistance())
+    >>> output = triplet_loss(anchor, positive, negative)
+    >>> output.backward()
+    >>>
+    >>> # Built-in Similarity Function
+    >>> triplet_loss = nn.TripletMarginLossWithDistance(distance_function=nn.CosineSimilarity(), is_similarity_function=True)
+    >>> output = triplet_loss(anchor, positive, negative)
+    >>> output.backward()
+    >>>
+    >>> # User-defined Similarity Function
+    >>> def l_infinity(x1, x2):
+    >>>     return torch.max(torch.abs(x1 - x2), dim=1).values
+    >>>
+    >>> triplet_loss = nn.TripletMarginLossWithDistance(distance_function=l_infinity, margin=1.5)
+    >>> output = triplet_loss(anchor, positive, negative)
+    >>> output.backward()
+
+    Reference:
+        V. Balntas, et al.: Learning shallow convolutional feature descriptors with triplet losses:
+        http://www.bmva.org/bmvc/2016/papers/paper119/index.html
+    """
+    __constants__ = ['is_similarity_function', 'margin', 'swap', 'reduction']
+    is_similarity_function: bool
+    margin: float
+    swap: bool
+
+    def __init__(self, distance_function: Optional[Callable[[Tensor, Tensor], Tensor]] = None, is_similarity_function: bool = False,
+                 margin: float = 1.0, swap: bool = False, reduction: str = 'mean'):
+        super(TripletMarginLossWithDistance, self).__init__(size_average=None, reduce=None, reduction=reduction)
+        self.distance_function = distance_function if distance_function is not None else PairwiseDistance()
+        self.is_similarity_function = is_similarity_function
+        self.margin = margin
+        self.swap = swap
+
+    def forward(self, anchor: Tensor, positive: Tensor, negative: Tensor) -> Tensor:
+        if not torch.jit.is_scripting():
```

Comment:
Maybe we should just give up on making this scriptable for the moment to avoid this code duplication. @albanD, thoughts?

Reply:
I am definitely not fond of the code duplication, especially because it does the same thing in both cases!
If you call the functional version with the distance function that comes from this module it does work right?
In particular, if I call the functional version with a function that is scriptable it should work no?  Should we just relax the constraint on the functional version?

Reply:
@albanD unfortunately, calling the functional version doesn't work -- I think JIT doesn't like that the function wants a Callable:
```
>>> loss_fn = torch.nn.TripletMarginLossWithDistance(distance_function=torch.nn.PairwiseDistance())
>>> torch.jit.script(loss_fn)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/users/echau18/pytorch/torch/jit/_script.py", line 887, in script
    return torch.jit._recursive.create_script_module(
  File "/data/users/echau18/pytorch/torch/jit/_recursive.py", line 317, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/data/users/echau18/pytorch/torch/jit/_recursive.py", line 376, in create_script_module_impl
    create_methods_from_stubs(concrete_type, stubs)
  File "/data/users/echau18/pytorch/torch/jit/_recursive.py", line 292, in create_methods_from_stubs
    concrete_type._create_methods(defs, rcbs, defaults)
  File "/data/users/echau18/pytorch/torch/jit/_recursive.py", line 596, in try_compile_fn
    return torch.jit.script(fn, _rcb=rcb)
  File "/data/users/echau18/pytorch/torch/jit/_script.py", line 929, in script
    fn = torch._C._jit_script_compile(
RuntimeError:
Unknown type constructor Callable:
# type: (Tensor, Tensor, Tensor, Optional[Callable[[Tensor, Tensor], Tensor]], bool, float, bool, str) -> Tensor
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
'triplet_margin_loss_with_distance' is being compiled since it was called from 'TripletMarginLossWithDistance.forward'
  File "/data/users/echau18/pytorch/torch/nn/modules/loss.py", line 1364
    def forward(self, anchor: Tensor, positive: Tensor, negative: Tensor) -> Tensor:
        return F.triplet_margin_loss_with_distance(anchor, positive, negative,
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                    distance_function=self.distance_function,
                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                    is_similarity_function=self.is_similarity_function,
                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                    margin=self.margin, swap=self.swap, reduction=self.reduction)
                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
```

Reply:
Ho right, it does not support functions that take Callable as arguments.
I would lean towards dropping the script support for now yes in that case.

Reply:
Do you mean dropping script support for both module and functional?  If so, is there a good way to denote that other than the error message I had?

Reply:
Yes, we do mean dropping script support for both. I think the error message is the best way to communicate that lack of support.